cranelift/codegen/meta/src/cdsl/typevar.rs

```diff
@@ -124,6 +124,14 @@ impl TypeVar {
         &self.type_set
     }
 
+    /// Returns true if this type variable refines but doesn't fully determine its value from the
+    /// parent type variable.
+    pub fn refines_parent(&self) -> bool {
+        self.base
+            .as_ref()
+            .map_or(false, |base| !base.derived_func.is_singleton())
+    }
```

Comment:
This function doesn't capture the case where a variable depends on a parent that refines yet another parent. It might be worth thinking a bit about how this would work, though we don't currently have any instructions that combine constraints like that.